camel/tasks/task.py

```diff
@@ -0,0 +1,281 @@
+    """
+    Task is specific assignment that can be passed to a agent.
+
+    Attributes:
+        content: string content for task.
+        id: A unique id.
+        state: The state which should be OPEN, RUNNING, DONE or DELETED.
+        parent: The parent task, None for root task.
+        subtasks: The childrent sub-tasks for the task.
+        result: The answer for the task.
+    """
+
+    content: str
+    """The content of the task."""
+
+    id: str
+    """An unique string identifier for the task. This should ideally be
+    provided by the provider/model which created the task."""
+
+    state: TaskSate = TaskSate.OPEN
+    """The task state.
+    """
+
+    parent: Optional["Task"] = None
+    """The parent task.
+    """
+
+    subtasks: List["Task"] = []
+    """A list of sub tasks.
+    """
+
+    result: str = ""
+
+    @classmethod
+    def from_message(cls, message: BaseMessage) -> "Task":
+        r"""Create a task from a message.
+
+        Args:
+            message (BaseMessage): The message to the task.
+
+        Returns:
+            Task
+        """
+        return cls(content=message.content, id="0")
+
+    @staticmethod
+    def to_message():
+        r"""Convert a Task to a Message."""
+        # TODO
+        pass
+
+    def reset(self):
+        r"""Reset Task to initial state."""
+        self.state = "OPEN"
+        self.result = ""
+
+    def update_task_result(self, message: BaseMessage):
+        self.result = message.content
+        self.state = TaskSate.OPEN
+
+    def set_id(self, id: str):
+        self.id = id
+
+    def set_state(self, state: TaskSate):
+        self.state = state
+        if state == TaskSate.DONE:
+            for subtask in self.subtasks:
+                if subtask.state != TaskSate.DELETED:
+                    subtask.set_state(state)
+        elif state == TaskSate.RUNNING:
+            if self.parent is not None:
+                self.parent.set_state(state)
+
+    def add_subtask(self, task: "Task"):
+        task.parent = self
+        self.subtasks.append(task)
+
+    def get_running_task(self) -> Optional["Task"]:
+        for sub in self.subtasks:
+            if sub.state == TaskSate.RUNNING:
+                return sub.get_running_task()
+        if self.state == TaskSate.RUNNING:
+            return self
+        return None
+
+    def to_string(self, indent="") -> str:
+        _str = f"{indent}Task {self.id}: {self.content}\n"
+        for subtask in self.subtasks:
+            _str += subtask.to_string(indent + "  ")
+        return _str
+
+
+def parse_response(response: str, task_id: Optional[str] = None) -> List[Task]:
+    pattern = "<task>(.*?)</task>"
+    tasks_content = re.findall(pattern, response, re.DOTALL)
+
+    tasks = []
+    if task_id is None:
+        task_id = "0"
+    for i, content in enumerate(tasks_content):
+        tasks.append(Task(content=content.strip(), id=f"{task_id}.{i}"))
+    return tasks
+
+
+class TaskManager:
```

Comment:
`TaskManager` seems to be designed for a single manager to coordinate all tasks. I was wondering because the `groupchat` or `workforce` process is recursive, do you think it's necessary to make this class recursive, so that if I choose one of these nodes it's not just a `task`, it's also a `TaskManager` that contains information like `task_map`. 

I'm not sure, what do you think are the advantages and disadvantages of accessing `TaskManager` recursively versus querying `TaskManager` by `id` of specific root node all the time?

Reply:
I think accessing TaskManager recursively versus querying TaskManager by id of specific root node can be implemented/achieved at the same time?

Reply:
Yes, the design of current `TaskManager` have already met the requirement of recursive call.

Reply:
Here the `TaskManager` is not a Agent.  I think `ask_agent` should refactored based on current Task design.This makes the structure clearer.

Reply:
Maybe it doesn't matter so much now. I will now deal directly with the results of decompose.